File: exercise04-B/src/program.cpp
// Exercise 04-B: Happy Number

#include <iostream>
#include <unordered_map>
#include <unordered_set>

// Prototypes

bool IsHappyCached(int n, std::unordered_map<int, bool>& cache);
int SquareDigitsSum(int n);

// Main execution

int main(int argc, char* argv[]) {
    std::unordered_map<int, bool> cache;

    int n;
    while (std::cin >> n)
        std::cout << (IsHappyCached(n, cache) ? "Yes" : "No") << '\n';

    return 0;
}

// Functions

bool IsHappyCached(int n, std::unordered_map<int, bool>& cache) {
    std::unordered_set<int> seen;
    bool is_happy = false;

    while (seen.find(n) == seen.end()) {
        if (cache.find(n) != cache.end()) {
            is_happy = cache[n];
            break;
        }

        seen.insert(n);
        n = SquareDigitsSum(n);
    }

    if (n == 1)
        is_happy = true;

    for (int s : seen) {
        cache[s] = is_happy;
    }

    return is_happy;
}

int SquareDigitsSum(int n) {
    int sum = 0;

    while (n > 0) {
        int digit = n % 10;
        sum += digit * digit;
        n /= 10;
    }

    return sum;
}

// vim: set sts=4 sw=4 ts=8 ft=cpp:

File: exercise04-B/src/program.py
#!/usr/bin/env python3

# Exercise 04-B: Happy Number

import sys

# Function


def is_happy_cached(n: int, cache: dict[int, bool]) -> bool:
    seen = set()
    happiness = False
    while n not in seen:
        if n in cache:
            happiness = cache[n]
            break

        seen.add(n)
        n = sum([x * x for x in [int(char) for char in str(n)]])

    if n == 1:
        happiness = True

    for s in seen:
        cache[s] = happiness

    return happiness


# Main execution


def main() -> None:
    cache = {}  # Memoization: keep track of numbers with known happiness
    for n in (int(line) for line in sys.stdin):
        print("Yes" if is_happy_cached(n, cache) else "No")


if __name__ == "__main__":
    main()

# vim: set sts=4 sw=4 ts=8 expandtab ft=python:
